api/src/main/java/org/pmiops/workbench/api/ConceptSetsController.java

```diff
@@ -0,0 +1,249 @@
+    dbConceptSet.setCreator(userProvider.get());
+    dbConceptSet.setWorkspaceId(workspace.getWorkspaceId());
+    dbConceptSet.setCreationTime(now);
+    dbConceptSet.setLastModifiedTime(now);
+    dbConceptSet.setVersion(1);
+    try {
+      dbConceptSet = conceptSetDao.save(dbConceptSet);
+      // TODO: add recent resource entry for concept sets?
+    } catch (DataIntegrityViolationException e) {
+      throw new BadRequestException(String.format(
+          "Concept set \"/%s/%s/%d\" already exists.",
+          workspaceNamespace, workspaceId, dbConceptSet.getName()));
+    }
+    return ResponseEntity.ok(toClientConceptSet(dbConceptSet));
+  }
+
+  private ConceptSet toClientConceptSet(org.pmiops.workbench.db.model.ConceptSet conceptSet) {
+    ConceptSet result = TO_CLIENT_CONCEPT_SET.apply(conceptSet);
+    if (!conceptSet.getConceptIds().isEmpty()) {
+      result.setConcepts(ImmutableList.copyOf(conceptDao.findAll(conceptSet.getConceptIds())).stream().map(
+            ConceptsController.TO_CLIENT_CONCEPT).sorted(CONCEPT_NAME_ORDERING).collect(Collectors.toList()));
+    }
+    return result;
+  }
+
+  @Override
+  public ResponseEntity<EmptyResponse> deleteConceptSet(String workspaceNamespace,
+      String workspaceId, Long conceptSetId) {
+    org.pmiops.workbench.db.model.ConceptSet conceptSet =
+        getDbConceptSet(workspaceNamespace, workspaceId, conceptSetId, WorkspaceAccessLevel.WRITER);
+    conceptSetDao.delete(conceptSet.getConceptSetId());
+    return ResponseEntity.ok(new EmptyResponse());
+  }
+
+  @Override
+  public ResponseEntity<ConceptSet> getConceptSet(String workspaceNamespace, String workspaceId,
+      Long conceptSetId) {
+    org.pmiops.workbench.db.model.ConceptSet conceptSet =
+        getDbConceptSet(workspaceNamespace, workspaceId, conceptSetId, WorkspaceAccessLevel.READER);
+    return ResponseEntity.ok(toClientConceptSet(conceptSet));
+  }
+
+  @Override
+  public ResponseEntity<ConceptSetListResponse> getConceptSetsInWorkspace(String workspaceNamespace,
+      String workspaceId) {
+    workspaceService.getWorkspaceEnforceAccessLevelAndSetCdrVersion(workspaceNamespace, workspaceId,
+        WorkspaceAccessLevel.READER);
+
+    Workspace workspace = workspaceService.getRequiredWithConceptSets(workspaceNamespace, workspaceId);
+    ConceptSetListResponse response = new ConceptSetListResponse();
+    Set<org.pmiops.workbench.db.model.ConceptSet> conceptSets = workspace.getConceptSets();
+    if (conceptSets != null) {
+      Set<Long> conceptIds = Sets.newHashSet();
+      // Get all the concept IDs for all concept sets, and look them all up at once.
+      for (org.pmiops.workbench.db.model.ConceptSet conceptSet : conceptSets) {
+        conceptIds.addAll(conceptSet.getConceptIds());
+      }
+      final Map<Long, Concept> conceptMap;
+      if (conceptIds.size() > 0) {
+        conceptMap = ImmutableList.copyOf(conceptDao.findAll(conceptIds)).stream()
```

Comment:
No -- this will actually return whatever concepts it finds, and ignore everything else.

If we want to ensure that there are no concept set IDs in here which aren't in the concept table of the corresponding CDR DB, we can add some business logic to enforce that; but either way I don't think we'd blow up here.

I am not sure how we're going to handle the case of pointing workspaces at a new CDR version, which in theory might not have concepts that were in the previous one... in those cases there's something to be said for copying the IDs over and just not showing them in the UI.